### DISCUSSION TITLE
Typing for a legacy `count` query:

### DISCUSSION THREAD

**Discussion by jvanasco** [sqlalchemy/sqlalchemy#11316] [Usage Questions] [answered]
I have a handful of legacy "count" queries:

    result = (
        dbSession.query(
            Foo.column,
            sqlalchemy.func.count(True).label("counted"),
        )
        .group_by(Foo.column)
        .all()
    )

These do not pass mypy with the following error:

    Argument 1 to "count" has incompatible type "bool"; expected "ColumnElement[Any] | _HasClauseElement[Any] | SQLCoreOperations[Any] | ExpressionElementRole[Any] | Callable[[], ColumnElement[Any]] | LambdaElement | None"

Is there a low-cost typing compliant way to restate this type of query?

I know I can just pass in the column by name.  Selecting `True` or `1` has often yielded faster results.

**Comment by CaselIT** [comment, marked as answer]
Hi,

The type could probably use an update in count. Easiest solution would be to use `count()` without any arg. That will generate either `count(1)` or `count(*)`

**Reply by jvanasco** [reply]
Thanks!  That works.  I just want to avoid the column name on this, as the planner shows better performance without it.

**Reply by CaselIT** [reply]
I've added https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5255

**Reply by zzzeek** [reply]
the gerrit above does not actually fix this case:

    func.count(True)

that's still a typing error.  it works for "1" and "*"

**Reply by CaselIT** [reply]
`count(True)` seems very strange to me. I never seen it before. Also I don't think it would work in all dbs (since some don't really have a boolean type).

So I would be to not support that in typing. But I'm kinda fine either way too

**Reply by zzzeek** [reply]
is `count(1)` still a thing?  I sort of thought even oracle had made `count('*')` the official way to do this

**Reply by jvanasco** [reply]
`1` and `*` are fine.  Sqlite casts `True` to `1` automatically for compat.   Most DBs I've seen support it, but `1` does the same thing and is good enough.

I have seen so many janky weird DB things in the past week while looking up some typing workarounds.  My favorite right now are the "tech bloggers" that recommend starting every multiple conditional with `WHERE 1 = 1`, so all subsequent conditionals can be formatted in an editor as:

    AND a = b
    
Why do that?  This way the sql still works if you comment out the first actual conditional you wrote:

    WHERE 1 = 1
    # AND a = b
    AND c = d

This is somehow now a recommended pattern.  I am very excited to see how this will affect poorly designed AI queries.

**Reply by CaselIT** [reply]
> is count(1) still a thing? I sort of thought even oracle had made count('*') the official way to do this

no clue, really, but I guess it may be popular since it's kinda similar to `select 1`

**Reply by CaselIT** [reply]
> My favorite right now are the "tech bloggers" that recommend starting every multiple conditional with WHERE 1 = 1, so all subsequent conditionals can be formatted in an editor as:

I think this is there since forever, especially in php. Half of wordpress query are like `query.replace('1=1', '1=1 and my_new_condition')`
